controller/rancher/certificate_fetcher.go

```diff
@@ -0,0 +1,317 @@
+package rancher
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/fsnotify/fsnotify"
+	"github.com/rancher/go-rancher-metadata/metadata"
+	"github.com/rancher/go-rancher/v2"
+	"github.com/rancher/lb-controller/config"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+const (
+	CertName = "fullchain.pem"
+	KeyName  = "privkey.pem"
+)
+
+type CertificateFetcher interface {
+	FetchCertificate(certName string) (*config.Certificate, error)
+	UpdateEndpoints(lbSvc *metadata.Service, eps []client.PublicEndpoint) error
+	ReadAllCertificatesFromDir(certDir string) ([]*config.Certificate, error)
+	ReadDefaultCertificate(defaultCertDir string) (*config.Certificate, error)
+	StopWatcher() error
+	ProcessFileUpdateEvents(do func(string))
+}
+
+type RCertificateFetcher struct {
+	Client         *client.RancherClient
+	CertsCache     map[string]*config.Certificate //cert name (sub dir name) -> cert
+	CertDir        string
+	DefaultCertDir string
+	DefaultCert    *config.Certificate
+	Watcher        *fsnotify.Watcher
+	mu             *sync.RWMutex
+	TempCertsMap   map[string]*config.Certificate //cert name (sub dir name) -> cert
```

Comment:
filepath.Walk method signature cannot be changed to pass any temp structure, it can only save the certs in some field that is part of RCertificateFetcher.